File: libraries/typescript/src/tts/sarvam.ts
/** Sarvam AI TTS for Patter pipeline mode. */
import {
  SarvamTTS as _SarvamTTS,
  SarvamAudioCodec,
  SarvamSampleRate,
  type SarvamLanguage,
  type SarvamModel,
} from '../providers/sarvam-tts';

/** Constructor options for the Sarvam `TTS` adapter. */
export interface SarvamTTSOptions {
  /** API key. Falls back to SARVAM_API_KEY env var when omitted. */
  apiKey?: string;
  model?: SarvamModel | string;
  speaker?: string;
  /** BCP-47 language code, e.g. `"hi-IN"`, `"ta-IN"`, `"en-IN"`. Default `"en-IN"`. */
  language?: SarvamLanguage | string;
  /** Audio codec. Default `linear16`; `mulaw` @ 8 kHz = carrier-native. */
  codec?: SarvamAudioCodec | string;
  sampleRate?: SarvamSampleRate | number;
  pace?: number;
  pitch?: number;
  loudness?: number;
  temperature?: number;
  enablePreprocessing?: boolean;
  dictId?: string;
  baseUrl?: string;
}

/** Options for the carrier-specific factories — same as the constructor minus `codec`/`sampleRate`. */
export type SarvamCarrierOptions = Omit<SarvamTTSOptions, 'codec' | 'sampleRate'>;

function resolveApiKey(apiKey: string | undefined): string {
  const key = apiKey ?? process.env.SARVAM_API_KEY;
  if (!key) {
    throw new Error(
      "Sarvam TTS requires an apiKey. Pass { apiKey: '...' } or " +
        'set SARVAM_API_KEY in the environment.',
    );
  }
  return key;
}

/**
 * Sarvam Bulbul TTS for Indian languages — defaults to the bulbul:v3 model.
 *
 * Synthesizes Hindi, Bengali, Tamil, Telugu, Kannada, Malayalam, Marathi,
 * Gujarati, Punjabi, Odia and Indian English (plus code-mixed text). Select the
 * language via `language` (a Sarvam BCP-47 code such as `"hi-IN"`).
 *
 * @example
 * ```ts
 * import * as sarvam from "getpatter/tts/sarvam";
 * const tts = new sarvam.TTS();                               // reads SARVAM_API_KEY
 * const tts = new sarvam.TTS({ language: "hi-IN", speaker: "shubh" });
 * ```
 *
 * **Telephony** — use {@link TTS.forTwilio} or {@link TTS.forTelnyx} on phone
 * calls. Both emit μ-law @ 8 kHz natively (Sarvam supports it), so the pipeline
 * skips resampling and PCM → μ-law encoding entirely (bit-clean passthrough).
 */
export class TTS extends _SarvamTTS {
  static readonly providerKey = 'sarvam';
  constructor(opts: SarvamTTSOptions = {}) {
    const key = resolveApiKey(opts.apiKey);
    const { apiKey: _ignored, ...rest } = opts;
    void _ignored;
    super(key, rest);
  }

  /** Pipeline TTS pre-configured for Twilio Media Streams (μ-law @ 8 kHz native — carrier-wire passthrough). */
  static override forTwilio(opts?: SarvamCarrierOptions): TTS;
  // Parent-compatible overload — accepts the legacy positional form too.
  static override forTwilio(apiKey: string, options?: SarvamCarrierOptions): TTS;
  static override forTwilio(
    arg1?: string | SarvamCarrierOptions,
    arg2?: SarvamCarrierOptions,
  ): TTS {
    const opts: SarvamCarrierOptions =
      typeof arg1 === 'string' ? { apiKey: arg1, ...(arg2 ?? {}) } : (arg1 ?? {});
    return new TTS({
      ...opts,
      codec: SarvamAudioCodec.MULAW,
      sampleRate: SarvamSampleRate.HZ_8000,
    });
  }

  /** Pipeline TTS pre-configured for Telnyx (μ-law @ 8 kHz native — PCMU wire passthrough). */
  static override forTelnyx(opts?: SarvamCarrierOptions): TTS;
  static override forTelnyx(apiKey: string, options?: SarvamCarrierOptions): TTS;
  static override forTelnyx(
    arg1?: string | SarvamCarrierOptions,
    arg2?: SarvamCarrierOptions,
  ): TTS {
    const opts: SarvamCarrierOptions =
      typeof arg1 === 'string' ? { apiKey: arg1, ...(arg2 ?? {}) } : (arg1 ?? {});
    return new TTS({
      ...opts,
      codec: SarvamAudioCodec.MULAW,
      sampleRate: SarvamSampleRate.HZ_8000,
    });
  }
}

File: libraries/python/getpatter/tts/sarvam.py
"""Sarvam AI TTS for Patter pipeline mode."""

from __future__ import annotations

import os
from typing import ClassVar, Optional, Union

from getpatter.providers.sarvam_tts import (
    SarvamAudioCodec,
    SarvamLanguage,
    SarvamModel,
    SarvamSampleRate,
    SarvamTTS as _SarvamTTS,
)

__all__ = ["TTS"]


class TTS(_SarvamTTS):
    """Sarvam Bulbul TTS for Indian languages — defaults to the bulbul:v3 model.

    Synthesizes Hindi, Bengali, Tamil, Telugu, Kannada, Malayalam, Marathi,
    Gujarati, Punjabi, Odia and Indian English (plus code-mixed text). Select
    the language via ``language`` (a Sarvam BCP-47 code such as ``"hi-IN"`` —
    see :class:`~getpatter.providers.sarvam_tts.SarvamLanguage`).

    Example::

        from getpatter.tts import sarvam

        tts = sarvam.TTS()                                  # reads SARVAM_API_KEY
        tts = sarvam.TTS(api_key="...", language="hi-IN", speaker="shubh")

        # Telephony (mu-law @ 8 kHz, carrier-native passthrough):
        tts = sarvam.TTS.for_twilio(language="ta-IN")
    """

    provider_key: ClassVar[str] = "sarvam"

    def __init__(
        self,
        api_key: Optional[str] = None,
        *,
        model: Union[SarvamModel, str] = SarvamModel.BULBUL_V3,
        speaker: str = "shubh",
        language: Union[SarvamLanguage, str] = SarvamLanguage.ENGLISH,
        codec: Union[SarvamAudioCodec, str] = SarvamAudioCodec.LINEAR16,
        sample_rate: Union[SarvamSampleRate, int] = SarvamSampleRate.HZ_16000,
        pace: Optional[float] = None,
        pitch: Optional[float] = None,
        loudness: Optional[float] = None,
        temperature: Optional[float] = None,
        enable_preprocessing: Optional[bool] = None,
        dict_id: Optional[str] = None,
    ) -> None:
        key = api_key or os.environ.get("SARVAM_API_KEY")
        if not key:
            raise ValueError(
                "Sarvam TTS requires an api_key. Pass api_key='...' or "
                "set SARVAM_API_KEY in the environment."
            )
        super().__init__(
            api_key=key,
            model=model,
            speaker=speaker,
            language=language,
            codec=codec,
            sample_rate=sample_rate,
            pace=pace,
            pitch=pitch,
            loudness=loudness,
            temperature=temperature,
            enable_preprocessing=enable_preprocessing,
            dict_id=dict_id,
        )
